File: Scripts/2021/euler_rotation_inverse.m
function [phi, theta, psi] = euler_rotation_inverse(sequence, R)
% [phi, theta, psi] = euler_rotation_inverse(sequence, R) takes as inputs:
%   -sequence: a string which specifies how the euler-rotation has been computed, e.g. "xyx"
%   -R: the rotation to be decomposed, should be a 3x3 matrix
% and outputs:
%   -phi: the radiants of the first rotation
%   -theta: the radiants of the second rotation
%   -psi: the radiants of the third rotation
% Upon execution the function will ask for an input which will determine
% which solution must be displayed (there is always a pair of solutions)
%
% Euler rotations work about moving-axes
    
    if isa(R, 'sym')
        disp("R must by non-symbolic matrix")
        phi = -1; theta = -1; psi = -1;
        return
    end
    
    if strlength(sequence) ~= 3
        disp("Sequence not valid, must be of length three.")
        return;
    end
    
    if (sequence(2) == sequence(1) || sequence(2) == sequence(3))
        disp("Two consecutive rotation along the same axis are not valid.")
        return
    end
    
    risp = input("There is always a pair of solutions to the inverse euler problem. Do you want to use the positive or negative sin(theta)? (pos, neg)\n", "s");
    cond = risp == "pos";
    
    switch lower(sequence)
        case "xyx"
            theta = atan2(sqrt(R(1, 2)^2 + R(1, 3)^2), R(1, 1))*cond + atan2(-sqrt(R(1, 2)^2 + R(1, 3)^2), R(1, 1))*(1-cond);
            if (abs(sin(theta)) <= 1e-6)
                disp('Singular case: sin(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(1, 2)/sin(theta), R(1, 3)/sin(theta));
            phi = atan2(R(2, 1)/sin(theta), -R(3, 1)/sin(theta));
            
        case "xyz"
            theta = atan2(R(1, 3), sqrt(R(1, 1)^2 + R(1, 2)^2))*cond + atan2(R(1, 3), -sqrt(R(1, 1)^2 + R(1, 2)^2))*(1-cond);
            if (abs(cos(theta)) <= 1e-6)
                disp('Singular case: cos(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(-R(1, 2)/cos(theta), R(1, 1)/cos(theta));
            phi = atan2(-R(2, 3)/cos(theta), R(3, 3)/cos(theta));
            
        case "xzx"
            theta = atan2(sqrt(R(1, 2)^2 + R(1, 3)^2), R(1, 1))*cond + atan2(-sqrt(R(1, 2)^2 + R(1, 3)^2), R(1, 1))*(1-cond);
            if (abs(sin(theta)) <= 1e-6)
                disp('Singular case: sin(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(1, 3)/sin(theta), -R(1, 2)/sin(theta));
            phi = atan2(R(3, 1)/sin(theta), R(2, 1)/sin(theta));
            
        case "xzy"
            theta = atan2(-R(1, 2), sqrt(R(1, 1)^2 + R(1, 3)^2))*cond + atan2(-R(1, 1), -sqrt(R(1, 3)^2 + R(2, 1)^2))*(1-cond);
            if (abs(cos(theta)) <= 1e-6)
                disp('Singular case: cos(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(1, 3)/cos(theta), R(1, 1)/cos(theta));
            phi = atan2(R(3, 2)/cos(theta), R(2, 2)/cos(theta));
            
        case "yxy"
            theta = atan2(sqrt(R(2, 3)^2 + R(2, 1)^2), R(2, 2))*cond + atan2(-sqrt(R(2, 3)^2 + R(2, 1)^2), R(2, 2))*(1-cond);
            if (abs(sin(theta)) <= 1e-6)
                disp('Singular case: sin(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(2, 1)/sin(theta), -R(2, 3)/sin(theta));
            phi = atan2(R(1, 2)/sin(theta), R(3, 2)/sin(theta));
            
        case "yxz"
            theta = atan2(-R(2, 3), sqrt(R(2, 2)^2 + R(2, 1)^2))*cond + atan2(-R(2, 3), -sqrt(R(2, 2)^2 + R(2, 1)^2))*(1-cond);
            if (abs(cos(theta)) <= 1e-6)
                disp('Singular case: cos(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(2, 1)/cos(theta), R(2, 2)/cos(theta));
            phi = atan2(R(1, 3)/cos(theta), R(3, 3)/cos(theta));
            
        case "yzx"
            theta = atan2(R(2, 1), sqrt(R(2, 2)^2 + R(2, 3)^2))*cond + atan2(R(2, 1), -sqrt(R(2, 2)^2 + R(2, 3)^2))*(1-cond);

            if (abs(cos(theta)) <= 1e-6)
                disp('Singular case: cos(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(-R(2, 3)/cos(theta), R(2, 2)/cos(theta));
            phi = atan2(-R(3, 1)/cos(theta), R(1, 1)/cos(theta));
            
        case "yzy"
            theta = atan2(sqrt(R(2, 1)^2 + R(2,3)^2), R(2, 2))*cond + atan2(-sqrt(R(2, 1)^2 + R(2,3)^2), R(2, 2))*(1-cond);
            if (abs(sin(theta)) <= 1e-6)
                disp('Singular case: sin(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(2, 3)/sin(theta), R(2, 1)/sin(theta));
            phi = atan2(R(3, 2)/sin(theta), -R(1, 2)/sin(theta));
            
        case "zxy"
            theta = atan2(R(3, 2), sqrt(R(3, 1)^2 + R(3,3)^2))*cond + atan2(R(3, 2), -sqrt(R(3, 1)^2 + R(3,3)^2))*(1-cond);
            if (abs(cos(theta)) <= 1e-6)
                disp('Singular case: cos(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(-R(3, 1)/cos(theta), R(3, 3)/cos(theta));
            phi = atan2(-R(1, 2)/cos(theta), R(2, 1)/cos(theta));
            
        case "zxz"
            theta = atan2(sqrt(R(1, 3)^2 + R(2, 3)^2), R(3, 3))*cond + atan2(-sqrt(R(1, 3)^2 + R(2, 3)^2), R(3, 3))*(1-cond);
            if (abs(sin(theta)) <= 1e-6)
                disp('Singular case: sin(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(3, 1)/sin(theta), R(3, 2)/sin(theta));
            phi = atan2(R(1, 3)/sin(theta), -R(2, 3)/sin(theta));
            
        case "zyx"
            theta = atan2(-R(3, 1), sqrt(R(3, 2)^2+R(3, 3)^2))*cond + atan2(-R(3, 1), -sqrt(R(3, 2)^2+R(3, 3)^2))*(1-cond);
            if (abs(cos(theta)) <= 1e-6)
                disp('Singular case: cos(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(3, 2)/cos(theta), R(3, 3)/cos(theta));
            phi = atan2(R(2, 1)/cos(theta), R(1, 1)/cos(theta));
            
        case "zyz"
            theta = atan2(sqrt(R(3, 1)^2 + R(3, 2)^2), R(3, 3))*cond + atan2(-sqrt(R(3, 1)^2 + R(3, 2)^2), R(3, 3))*(1-cond);
            if (abs(sin(theta)) <= 1e-6)
                disp('Singular case: sin(theta) == 0 or very close to 0.')
                return
            end
            
            psi = atan2(R(3, 2)/sin(theta), -R(3, 1)/sin(theta));
            phi = atan2(R(2, 3)/sin(theta), R(1, 3)/sin(theta));
            
        otherwise
            disp("Invalid sequence")
    end
    
end

File: Scripts/2025/python_scripts/robobox/euler_rotation_inverse.py
import numpy as np

def euler_rotation_inverse(sequence: str, R: np.ndarray, solution_sign: str) -> list:
    """
    Computes the Inverse Euler rotation
    
    @param: sequence: The rotation sequence (e.g., "xyx", "xyz", etc.)
    @param: R: The 3x3 rotation matrix to be decomposed
    @param: solution_sign: Determines which solution to use ("pos" or "neg")
    """
    if R.shape != (3, 3):
        raise ValueError("R must be a 3x3 matrix.")
    
    if len(sequence) != 3:
        raise ValueError("Sequence must be of length 3.")
    
    if sequence[1] == sequence[0] or sequence[1] == sequence[2]:
        raise ValueError("Two consecutive rotations along the same axis are not valid.")
    
    cond = solution_sign == "pos"
    
    if sequence.lower() == "xyx":
        theta = np.arctan2(np.sqrt(R[0, 1]**2 + R[0, 2]**2), R[0, 0])*cond + \
                np.arctan2(-np.sqrt(R[0, 1]**2 + R[0, 2]**2), R[0, 0])*(1-cond)
        if np.abs(np.sin(theta)) <= 1e-6:
            print("Singular case: sin(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(R[0, 1]/np.sin(theta), R[0, 2]/np.sin(theta))
        phi = np.arctan2(R[1, 0]/np.sin(theta), -R[2, 0]/np.sin(theta))
    
    elif sequence.lower() == "xyz":
        theta = np.arctan2(R[0, 2], np.sqrt(R[0, 0]**2 + R[0, 1]**2))*cond + \
                 np.arctan2(R[0, 2], -np.sqrt(R[0, 0]**2 + R[0, 1]**2))*(1-cond)
        if np.abs(np.cos(theta)) <= 1e-6:
            print("Singular case: cos(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(-R[0, 1]/np.cos(theta), R[0, 0]/np.cos(theta))
        phi = np.arctan2(-R[1, 2]/np.cos(theta), R[2, 2]/np.cos(theta))
    
    elif sequence.lower() == "xzx":
        theta = np.arctan2(np.sqrt(R[0, 1]**2 + R[0, 2]**2), R[0, 0])*cond + \
                 np.arctan2(-np.sqrt(R[0, 1]**2 + R[0, 2]**2), R[0, 0])*(1-cond)
        if np.abs(np.sin(theta)) <= 1e-6:
            print("Singular case: sin(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(R[0, 2] / np.sin(theta), -R[0, 1] / np.sin(theta))
        phi = np.arctan2(R[2, 0] / np.sin(theta), R[1, 0] / np.sin(theta))
    
    elif sequence.lower() == "xzy":
        theta = np.arctan2(-R[0, 1], np.sqrt(R[0, 0]**2 + R[0, 2]**2))*cond + \
                 np.arctan2(-R[0, 0], -np.sqrt(R[0, 2]**2 + R[1, 0]**2))*(1-cond)
        if np.abs(np.cos(theta)) <= 1e-6:
            print("Singular case: cos(theta) ==0 or very close to 0.")
            return
        psi = np.arctan2(R[0, 2] / np.cos(theta), R[0, 0] / np.cos(theta))
        phi = np.arctan2(R[2, 1] / np.cos(theta), R[1, 1] / np.cos(theta))
    
    elif sequence.lower() == "yxy":
        theta = np.arctan2(np.sqrt(R[1, 2]**2 + R[1, 0]**2), R[1, 1])*cond + \
                 np.arctan2(-np.sqrt(R[1, 2]**2 + R[1, 0]**2), R[1, 1])*(1-cond)
        if np.abs(np.sin(theta)) <= 1e-6:
            print("Singular case: sin(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(R[1, 0] / np.sin(theta), -R[1, 2] / np.sin(theta))
        phi = np.arctan2(R[0, 1] / np.sin(theta), R[2, 1] / np.sin(theta))
    
    elif sequence.lower() == "yxz":
        theta = np.arctan2(-R[1, 2], np.sqrt(R[1, 1]**2 + R[1, 0]**2))*cond + \
                 np.arctan2(-R[1, 2], -np.sqrt(R[1, 1]**2 + R[1, 0]**2))*(1-cond)
        if np.abs(np.cos(theta)) <= 1e-6:
            print("Singular case: cos(theta) ==0 or very close to 0.")
            return
        psi = np.arctan2(R[1, 0] / np.cos(theta), R[1, 1] / np.cos(theta))
        phi = np.arctan2(R[0, 2] / np.cos(theta), R[2, 2] / np.cos(theta))
    
    elif sequence.lower() == "yzx":
        theta = np.arctan2(R[1, 0], np.sqrt(R[1, 1]**2 + R[1, 2]**2))*cond + \
                 np.arctan2(R[1, 0], -np.sqrt(R[1, 1]**2 + R[1, 2]**2))*(1-cond)
        if np.abs(np.cos(theta)) <= 1e-6:
            print("Singular case: cos(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(-R[1, 2] / np.cos(theta), R[1, 1] / np.cos(theta))
        phi = np.arctan2(-R[2, 0] / np.cos(theta), R[0, 0] / np.cos(theta))
    
    elif sequence.lower() == "yzy":
        theta = np.arctan2(np.sqrt(R[1, 0]**2 + R[1, 2]**2), R[1, 1])*cond + \
                 np.arctan2(-np.sqrt(R[1, 0]**2 + R[1, 2]**2), R[1, 1])*(1-cond)
        if np.abs(np.sin(theta)) <= 1e-6:
            print("Singular case: sin(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(R[1, 2] / np.sin(theta), -R[1, 0] / np.sin(theta))
        phi = np.arctan2(R[2, 1] / np.sin(theta), -R[0, 1] / np.sin(theta))
    
    elif sequence.lower() == "zxy":
        theta = np.arctan2(R[2, 1], np.sqrt(R[2, 0]**2 + R[2, 2]**2))*cond + \
                 np.arctan2(R[2, 1], -np.sqrt(R[2, 0]**2 + R[2, 2]**2))*(1-cond)
        if np.abs(np.cos(theta)) <= 1e-6:
            print("Singular case: cos(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(-R[2, 0] / np.cos(theta), R[2, 2] / np.cos(theta))
        phi = np.arctan2(-R[0, 1] / np.cos(theta), R[1, 0] / np.cos(theta))
    
    elif sequence.lower() == "zxz":
        theta = np.arctan2(np.sqrt(R[0, 2]**2 + R[1, 2]**2), R[2, 2])*cond + \
                 np.arctan2(-np.sqrt(R[0, 2]**2 + R[1, 2]**2), R[2, 2])*(1-cond)
        if np.abs(np.sin(theta)) <= 1e-6:
            print("Singular case: sin(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(R[2, 0] / np.sin(theta), R[2, 1] / np.sin(theta))
        phi = np.arctan2(R[0, 2] / np.sin(theta), -R[1, 2] / np.sin(theta))
    
    elif sequence.lower() == "zyx":
        theta = np.arctan2(-R[2, 0], np.sqrt(R[2, 1]**2 + R[2, 2]**2))*cond + \
                 np.arctan2(-R[2, 0], -np.sqrt(R[2, 1]**2 + R[2, 2]**2))*(1-cond)
        if np.abs(np.cos(theta)) <= 1e-6:
            print("Singular case: cos(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(R[2, 1] / np.cos(theta), R[2, 2] / np.cos(theta))
        phi = np.arctan2(R[1, 0] / np.cos(theta), R[0, 0] / np.cos(theta))
    
    elif sequence.lower() == "zyz":
        theta = np.arctan2(np.sqrt(R[2, 0]**2 + R[2, 1]**2), R[2, 2])*cond + \
                 np.arctan2(-np.sqrt(R[2, 0]**2 + R[2, 1]**2), R[2, 2])*(1-cond)
        if np.abs(np.sin(theta)) <= 1e-6:
            print("Singular case: sin(theta) = 0 or very close to 0.")
            return
        psi = np.arctan2(R[2, 1] / np.sin(theta), -R[2, 0] / np.sin(theta))
        phi = np.arctan2(R[1, 2] / np.sin(theta), R[0, 2] / np.sin(theta))
    else:
        raise ValueError("Invalid sequence.")
    
    return phi, theta, psi
